codec/src/main/java/io/netty/handler/codec/compression/ZstdDecoder.java

```diff
@@ -0,0 +1,186 @@
+        this(DEFAULT_MAX_BLOCK_SIZE);
+    }
+
+    /**
+     * Creates a new Zstd decoder.
+     *  @param  maxBlockSize
+     *            specifies the max block size
+     */
+    public ZstdDecoder(int maxBlockSize) {
+        this.maxBlockSize = ObjectUtil.checkPositive(maxBlockSize, "maxBlockSize");
+    }
+
+    /**
+     * Current state of stream.
+     */
+    private enum State {
+        DECOMPRESS_DATA,
+        NEED_MORE_DATA,
+        FINISHED,
+        CORRUPTED
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        try {
+            while (in.isReadable()) {
+                switch (currentState) {
+                    case DECOMPRESS_DATA:
+                        decompressData(ctx, in, out);
+                        break;
+                    case NEED_MORE_DATA:
+                        buffer.writeBytes(in);
+                        decompressData(ctx, buffer, out);
+                        break;
+                    case FINISHED:
+                    case CORRUPTED:
+                        in.skipBytes(in.readableBytes());
+                        break;
+                    default:
+                        throw new IllegalStateException();
+                }
+            }
+        } catch (Exception e) {
+            currentState = State.CORRUPTED;
+            throw e;
+        }
+    }
+
+    private boolean consumeAndDecompress(ChannelHandlerContext ctx, int decompressedSize,
+                                         ByteBuf in, List<Object> out) throws IOException {
+        // Bytebuf cannot release here because ByteToMessageDecoder will release it
+        is = new ByteBufInputStream(in, false);
+        zstdIs = new ZstdInputStreamNoFinalizer(is);
+        // setContinuous to true so that ZstdInputStreamNoFinalizer.read() will
+        // return -1 when decompression is not completed
+        zstdIs.setContinuous(true);
+        bos = new ByteArrayOutputStream();
+        IOUtils.copy(zstdIs, bos);
+        byte[] decompressed = bos.toByteArray();
+        int decompressedLength = decompressed.length;
+        // Check the decompression status since we use ZstdInputStream as an indicator
+        // to determine whether the decompression is completed
+        // TODO: Use Zstd to check the compressed data when zstd library support this
+        if (decompressedLength == 0 || (decompressedSize > 0 && decompressedSize != decompressedLength)) {
+            in.readerIndex(0);
+            if (currentState == State.DECOMPRESS_DATA) {
+                if (buffer == null) {
+                    if (RECOMMENDED_OUT_SIZE < Integer.MIN_VALUE || RECOMMENDED_OUT_SIZE > Integer.MAX_VALUE) {
+                        throw new IllegalStateException("Invalid recommendedOutSize");
+                    }
+                    buffer = ctx.alloc().buffer((int) RECOMMENDED_OUT_SIZE);
+                }
+                buffer.writeBytes(in);
```

Comment:
Why do we need this extra `buffer` at all ? Seems like we should be able to just use `in` directly without doing an extra memory cop.y

Reply:
The `buffer` here is an accumulator. When the data is transmitted in segments, the data read from the network will be stored in `in` each time, and then we will add it to the accumulator and decompress it. At this time, the data in `in` It is the next data block, so `buffer` is needed to store the last read data

Reply:
But why do we need this at all ? I mean `ByteToMessageDecoder` should just accumulate itself if there is not enough data to make progress and try again later.

Reply:
I'm not sure where I missed
`Zstd` does not support multi-block compression. When decompressing, `Zstd` will read the block header first. If the compressed data block does not have a block header, `Zstd` will throw `com.github.luben.zstd.ZstdIOException: Unknown frame`. When calling `decode()`, `in` is the currently arriving data block, which is already a segmented data block at this time, so if this data block does not have a block header, `ZstdDecoder` will not work normally, and `ByteToMessageDecoder` will release the `in` when decode() is called, so the accumulator is needed

Reply:
I am still confused... `ByteToMessageDecoder` might only release something when you read it, so if you not read it and so not adjust the `readerIndex` it will not release anything and just try again later when more data arrives.

Reply:
But `Zstd` must read the data of `in`. Do you mean to record the `readerIndex` and `writerIndex` before reading and reset when the compressed data block is not a complete block? I am not sure that this will bring better performance, but it does It's a bit of an odd idea to be honest.

Reply:
yeah basically just save the current readerIndex and reset it to the original position if you need the data again. This will save some allocations and memory copies

Reply:
Even if you do accumulation in `ByteToMessageDecoder`, you still need to write bytes to `in`, and writing bytes in `ZstdDecoder` does not seem to increase the copy operation, the only advantage is it can save a `buffer` space for decompression

Reply:
it saves allocations... `ByteToMessageDecoder` has all what you need. It's just odd to add an extra buffer that you also need to keep track of etc.

Reply:
<img width="1287" alt="image" src="https://github.com/netty/netty/assets/24522024/b8c17b20-df8b-4200-9377-4a36cecb2143">
I'm running into weird issues. When I reset the `readerIndex` and `writerIndex` in decompression, when the next time `decode()` is called, the `writerIndex` of `in` remains unchanged


```
private boolean consumeAndDecompress(ChannelHandlerContext ctx, int decompressedSize,
                                         ByteBuf in, List<Object> out) throws IOException {
        in.markReaderIndex();
        in.markWriterIndex();
       
        if (decompressedLength == 0 || (decompressedSize > 0 && decompressedSize != decompressedLength)) {
            in.resetReaderIndex();
            in.resetWriterIndex();
          
            return false;
        }
        final ByteBuf uncompressed = Unpooled.wrappedBuffer(decompressed);
        out.add(uncompressed);
        return true;
    }
```

Reply:
Let me have a look later today

Reply:
Done